### PR TITLE
Refactor: Carp Infusion & Lungs

### LINKED ISSUE #72678: Space Carp Infusion does not prevent suffocation in space
## Reproduction:

Infused human with six carp, and had all of the space carp infusions. This included : Big jaw, fish gills, space swimming, and the nomadic debuffs. When the human goes into space they have the pressure immunity that comes with the infusions, but they started suffocating despite having gills, and proceeded to nearly die to oxygen damage. The human also could not use the "space swimming" part of the infusion, which is supposed to be a free internal jetpack. The oxygen damage would not regenerate once in a air rich tile, and required medicine to get rid of it. 

**This has been verified on a test server as of reporting this issue.**
1. Get 4-6 carp and infuse a human with them.
2. Verify that they have all infusions with the infusion organs.
3. Go into space.
4. Proceed to suffocate in space, proceed to be unable to swim in space.

### PR DESCRIPTION
**Pull request by Floofies**
## About The Pull Request
Recently before merging PR #71867 I failed to account for the carp lungs and spacewalk trait in my testing (I did not try to fly across space) and so I did not notice other bugs. This PR fixes the bugs/oversights which I missed in my prior PR. Fixes #72678

I have completed a refactor of `/obj/item/organ/internal/lungs/proc/check_breath` and `/mob/living/carbon/proc/check_breath` because I noticed that bugs within the gas alert system (stuck alerts) were caused by their improper execution, and the code was written in a hard-to-understand style.

This PR also includes a small adjustment to status indicators which allows you to add an already existing timed indicator without overwriting it; it instead resets the existing timeout.

Here is a granular list of bugs:

- Bug A: Causes the spacewalk trait to not be applied to the mob, and is caused by a minor developer oversight in `/datum/status_effect/organ_set_bonus/carp/enable_bonus()` wherein it attempts to add `TRAIT_SPACEWALK` to the status effect rather than to the Mob.
- Bug B: Causes the carp lungs to suffocate you in space, because that behavior is unimplemented.
- Bug C: Causes the gas alert indicators ("Too much O2!" etc) to stay stuck to the screen if `check_breath` or other procs fail to remove them. This is caused by the way `check_breath` early-returns without executing the necessary logic to manage its gas alerts.
- Bug D:  The part of `lungs/proc/check_breath()` that removes consumed miasma from the air is only doing so when helium is also present and consumed.

This PR contains fixes for these bugs:

- Fix A: Added the spacewalk trait to the mob, allowing the trait to enable space flight as expected.
- Fix B: Added `TRAIT_SPACEBREATHING` as a lungs trait, which is used in `check_breath`, which allows the mob to "breathe" in space or vacuum, but suffocate when breathing unsuitable gas/O2.
- Fix C: Changed the proc `/mob/proc/throw_alert` such that, when throwing an alert which already exists, it resets the timeout of the existing alert rather than replacing it.
  - ~Fix C2: Added a 10 second timer to gas alerts. If something fails to clear them from the screen, they will now clear themselves after 10 seconds.~
- Fix D: Refactored `/mob/living/carbon/proc/check_breath` and `obj/item/organ/internal/lungs/proc/check_breath` to run the necessary logic for resetting gas side effects and gas alerts.
  - Removed an early-return which was causing the `check_breath` functions to skip most of their critical sections.
  - Fixed the part of `lungs/proc/check_breath()` that removes consumed miasma from the air, allowing it to remove the correct amount when expected to.
  - Added conditionals to check for `TRAIT_SPACEBREATHING` on lungs to bypass suffocation.
  - Added a fall-back to a default "flyweight" breath `datum/gas_mixture/empty_breath` to Carbons and lungs, used in in the event that either gets a null `breath`.
  - Renamed `obj/item/organ/internal/lungs/proc/handle_too_little_breath` to `handle_suffocation` and refactored it to handle more aspects of suffocation that was originally being handled inline.
  - Added `/obj/item/organ/internal/lungs/proc/breathe_gas_volume` to `check_breath` to replace lots of inlined logic which handles consuming/exchanging gases (O2 for CO2, etc).

## Why It's Good For The Game

Shortly after PR #71867 was merged, we noticed the carp lungs and spacewalking trait were not working as expected when doing the Space Carp DNA infusion. This PR fixes the bug and allows you to spacewalk like a Space Carp again.

This PR also adjusts the gas alert indicators so they don't get stuck to the screen as often. This PR also fixes a bug that was causing timed alerts to overwrite each other when it was unnecessary to do so, and helps prevent their on-throw animation from being replayed when not appropriate.

I have included an intermediate refactor of `/mob/living/carbon/proc/check_breath` and `obj/item/organ/internal/lungs/proc/check_breath` in this PR, and I have tried to improve readability/maintainability while correcting critical oversights in the ways breathing was working. After my changes are merged, I will likely work on another refactor to delete duplicated sections of code between Carbons and lungs, as much of it is 99% identical.

## Changelog

:cl: A.C.M.O.
refactor: Refactored breathing, mostly check_breath, for Carbons and lungs organ.
add: Added the space-breathing trait for lungs.
fix: Fixed the Space Carp DNA infusion to apply the Spacewalk trait, allowing you to fly through space.
fix: Fixed the Space Carp lungs, allowing you to breathe in space with them.
fix: Fixed gas status indicators re-playing their on-throw animation when they are refreshed.
fix: Fixed gas status indicators getting stuck to the screen in some situations.
fix: Fixed gas side effects, such as euphoria and hallucinations, to reset when expected.
fix: For Humans, fixed Miasma only being consumed from the air when Helium is present.
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
Thank you!

**Comment by Floofies** [comment]
Okay. In this PR I have cleaned up all the oversights which my colleagues and I could detect. There was also an unimplemented behavior in the carp lungs which necessitated a custom `check_breath` solution.

I have now tested the changes OK, and this PR should be OK to merge!

**Comment by LemonInTheDark** [comment]
> Fix C2: Added a 10 second timer to gas alerts. If something fails to clear them from the screen, they will now clear themselves after 10 seconds.

I don't like this,
In case we delay hard enough that 10 seconds pass it will behave strangely, and I also do not like needing to update things on a loop like this, a hope of mine for a while now has been to make that one and done

**Comment by Floofies** [comment]
@LemonInTheDark Sure, I removed the timeouts. I am working on a workaround now by adding a `breathe_vacuum` variable to the lungs which allows `check_breath` to allow breathing in space without overriding it. I can not override it without causing gas alerts to become stuck to the screen.

**Comment by Floofies** [comment]
I have just completed a refactor in `lungs.dm` which allows its critical sections to run as expected, notably the areas which are concerned with gas alerts and breathing in a vacuum. I've tried to increase readability and maintainability in doing so!

**Comment by LemonInTheDark** [comment]
I am a bum

**Comment by LemonInTheDark** [comment]
swear I keep going to review this and then not

**Review by ShizCalev — CHANGES_REQUESTED** [review]
Make the can_breathe_vacuum var a trait instead. Refer to #72686

The issue also needs to be fixed in /mob/living/carbon/check_breath() as well, you only fixed it for humans.

File: code/game/machinery/dna_infuser/organ_sets/carp_organs.dm [hunk 1]
```diff
@@ -13,17 +13,21 @@
...
 
 ///Carp lungs! You can breathe in space! Oh... you can't breathe on the station, you need low oxygen environments.
+/// Inverts behavior of lungs. Bypasses suffocation due to space / lack of gas, but also allows Oxygen to suffocate.
 /obj/item/organ/internal/lungs/carp
 	name = "mutated carp-lungs"
 	desc = "Carp DNA infused into what was once some normal lungs."
-	safe_oxygen_min = 0 //we don't breathe this!
+	// Oxygen causes suffocation.
+	safe_oxygen_min = 0
+	safe_oxygen_max = 1
```
Comment (ShizCalev) [inline code/game/machinery/dna_infuser/organ_sets/carp_organs.dm:29]: Low O2 environments (enough to hurt humans) would start at 15 for the record. 

Up to you if you want to change it, but I personally feel like having the safe_oxygen_max being right at the line where humans would start suffocating due to low O2 levels would be more in line with the theme.

File: code/modules/mob/living/carbon/carbon_defines.dm [hunk 2]
```diff
@@ -58,6 +58,8 @@
 
 	///This is used to determine if the mob failed a breath. If they did fail a brath, they will attempt to breathe each tick, otherwise just once per 4 ticks.
 	var/failed_last_breath = FALSE
+	///Empty breath - Used in the event that [carbon/proc/check_breath()] gets a null breath.
+	var/datum/gas_mixture/empty_breath = new
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/carbon_defines.dm:62]: this seems wrong

File: code/modules/mob/living/carbon/life.dm [hunk 3]
```diff
@@ -130,145 +130,199 @@
...
 	if(HAS_TRAIT(src, TRAIT_NOBREATH))
-		return FALSE
+		return
 
-	var/obj/item/organ/internal/lungs = getorganslot(ORGAN_SLOT_LUNGS)
-	if(!lungs)
-		adjustOxyLoss(2)
+	// Breath may be null, so use a fallback "empty breath" for convenience.
+	if(!breath)
+		breath = empty_breath
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/life.dm:156]: how would we get null air?

File: code/modules/mob/living/carbon/life.dm [hunk 4]
```diff
@@ -130,145 +130,199 @@
...
 
-	//CRIT
-	if(!breath || (breath.total_moles() == 0) || !lungs)
-		if(reagents.has_reagent(/datum/reagent/medicine/epinephrine, needs_metabolizing = TRUE) && lungs)
-			return FALSE
-		adjustOxyLoss(1)
+	// Ensure gas volumes are present.
+	breath.assert_gases(/datum/gas/bz, /datum/gas/carbon_dioxide, /datum/gas/freon, /datum/gas/plasma, /datum/gas/pluoxium, /datum/gas/miasma, /datum/gas/nitrous_oxide, /datum/gas/nitrium, /datum/gas/oxygen)
+
+	/// The list of gases in the breath.
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/life.dm:161]: this seems like kinda redundant documentation

File: code/modules/surgery/organs/lungs.dm [hunk 5]
```diff
@@ -97,403 +99,519 @@
...
+/// Remove a volume of gas from the breath. Used to simulate absorbtion and interchange of gas in the lungs.
+/// Removes all of the given gas type unless given a volume argument.
+/// Returns the amount of gas theoretically removed.
+/obj/item/organ/internal/lungs/proc/breathe_gas_volume(list/breath_gases, datum/gas/remove_gas, datum/gas/exchange_gas = null, volume = null)
+	if(volume == null)
+		volume = breath_gases[remove_gas][MOLES]
+	breath_gases[remove_gas][MOLES] -= volume
+	if(exchange_gas)
+		breath_gases[exchange_gas][MOLES] += volume
+	return volume
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/surgery/organs/lungs.dm:592]: very slightly safer, plus null is cringe

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Is the purpose of this to ensure alerts are always cleared? I'm a bit unhappy with the current way we do breathing, mostly from a performance standpoint. We don't have a way to check what gas we last breathed, so we need to clear all non active alerts. it's a not insignificant cost to human life()

File: code/modules/mob/living/carbon/life.dm [hunk 6]
```diff
@@ -130,145 +130,199 @@
...
 	return FALSE
 
-
-//Third link in a breath chain, calls handle_breath_temperature()
+/**
+ * This proc tests if the lungs can breathe, if the mob can breathe a given gas mixture, and throws/clears gas alerts.
+ * If there are moles of gas in the given gas mixture, side-effects may be applied/removed on the mob.
+ * This proc expects a lungs organ in order to breathe successfully, but does not defer any work to it.
+ *
+ * Returns TRUE if the breath was successful, or FALSE if otherwise.
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:138]: It was existing behavior which I documented. I don't think it's used!

Right now it only sets to FALSE if the gas mixture is null or empty, and the mob can't breathe.

File: code/modules/mob/living/carbon/life.dm (hunk 3 above)
Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:156]: Space! As you showed me in Discord, `space/remove_air(amount)` returns `null`

File: code/modules/mob/living/carbon/life.dm (hunk 4 above)
Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:161]: True. I added that after a few folks questioned the usage of `assert_gases`

File: code/modules/mob/living/carbon/life.dm [hunk 7]
```diff
@@ -130,145 +130,199 @@
...
+	// Ensure gas volumes are present.
+	breath.assert_gases(/datum/gas/bz, /datum/gas/carbon_dioxide, /datum/gas/freon, /datum/gas/plasma, /datum/gas/pluoxium, /datum/gas/miasma, /datum/gas/nitrous_oxide, /datum/gas/nitrium, /datum/gas/oxygen)
+
+	/// The list of gases in the breath.
+	var/list/breath_gases = breath.gases
+	/// Indicates if there are moles of gas in the breath.
+	var/has_moles = breath.total_moles() != 0
 
+	// Validate lungs organ.
+	// FIXME: This is mostly original behavior, and it doesn't check organ damage.
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:167]: Whoops I left that there for myself and didn't mean to commit it. The comment is meant to indicate that the behavior of `carbon/check_breath`  is bypassing `lungs/check_breath` and that it isn't necessarily intentional.

File: code/modules/mob/living/carbon/life.dm [hunk 8]
```diff
@@ -130,145 +130,199 @@
...
+		pluoxium_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/pluoxium][MOLES])
+		o2_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/oxygen][MOLES] + (8 * pluoxium_pp))
+		plasma_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/plasma][MOLES])
+		co2_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/carbon_dioxide][MOLES])
+		// Partial pressures of "trace" gases.
+		bz_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/bz][MOLES])
+		freon_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/freon][MOLES])
+		miasma_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/miasma][MOLES])
+		n2o_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/nitrous_oxide][MOLES])
+		nitrium_pp = breath.get_breath_partial_pressure(breath_gases[/datum/gas/nitrium][MOLES])
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:220]: The purpose of the `if` is to reduce the amount of unnecessary overhead of the proc. Without any moles of gas, we can assume all those variables are zero, and only retrieve the partial pressures when needed.

File: code/modules/mob/living/carbon/life.dm [hunk 9]
```diff
@@ -130,145 +130,199 @@
...
-	//CARBON DIOXIDE
-	if(CO2_partialpressure > safe_co2_max)
+	//-- PLUOXIUM --//
+	// Behaves like Oxygen with 8X efficacy, but metabolizes into a reagent.
+	if(pluoxium_pp)
+		var/pluoxium_used = 0
+		// Inhale Pluoxium. Exhale nothing.
+		pluoxium_used = breath_gases[/datum/gas/pluoxium][MOLES]
+		breath_gases[/datum/gas/pluoxium][MOLES] -= pluoxium_used
+		// Metabolize to reagent.
```
[the remark this answers is not in the document]
Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:239]: Fixed by commit 1071f24ae577340af5936d98c05d9053659182df

File: code/modules/mob/living/carbon/carbon_defines.dm (hunk 2 above)
Comment (Floofies) [reply code/modules/mob/living/carbon/carbon_defines.dm]: Fixed by commit 1071f24ae577340af5936d98c05d9053659182df

File: code/modules/mob/living/carbon/life.dm [hunk 10]
```diff
@@ -130,145 +130,197 @@
...
 	if(HAS_TRAIT(src, TRAIT_NOBREATH))
-		return FALSE
+		return
 
-	var/obj/item/organ/internal/lungs = getorganslot(ORGAN_SLOT_LUNGS)
-	if(!lungs)
-		adjustOxyLoss(2)
+	// Breath may be null, so use a fallback "empty breath" for convenience.
+	if(!breath)
+		var/datum/gas_mixture/immutable/empty_breath = new(BREATH_VOLUME)
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/life.dm:156]: this can be static, so we don't need to recreate it

Comment (Floofies) [reply code/modules/mob/living/carbon/life.dm:156]: Oh sorry, I misinterpreted one of our earlier conversations! I will fix this

**Pull request merged**